ggml-alloc.c

```diff
@@ -57,7 +57,7 @@ struct free_block {
     size_t size;
 };
 
-#define MAX_FREE_BLOCKS 128
```

Comment:
During inference, the maximum number of free blocks that I have seen is only around 5. Maybe this much fragmentation is normal during training, but it may indicate a problem elsewhere.

Reply:
Hm, I guess there is more fragmentation during finetuning, because tensor lifetimes are longer than in inference. The forward activations need to be kept until they are used in backward pass. Gradient checkpointing considerably reduces this, hence resulting in less fragmention:

Without checkpointing, finetuning openllama-3b results in 112 free blocks, for openllama-7b results in 137 free blocks.
With checkpointing, finetuning openllama-3b results in 40 free blocks, for openllama-7b results in 42 free blocks.

But when there is enough RAM, it is useful to turn off checkpointing, because it requires an additional forward pass, for recomputing forward activations, resulting in increased runtime.

Reply:
The only issue is that `ggml_allocr_alloc_graph` may be quite slow with so many blacks, since `ggml_allocr_alloc` and `ggml_allocator_free_tensor` always check every free block. But if that becomes a problem, there is still a lot of room for optimization in `ggml-alloc`.

Reply:
This is not really a problem, the allocator is called only once before training and doesn't seem to take much time.

But less fragmentation would probably help further reducing the memory requirements.
Reordering the graph may help to achieve this, currently it is just topological ordered with depth first search by `ggml_visit_parents`. Maybe in the future this could improved using Kahn's algorithm and some heuristic to keep tensor lifetimes low.